march_hardware/src/EtherCAT/EthercatSDO.cpp

```diff
@@ -18,19 +18,36 @@ namespace march4cpp
 int sdo_bit8(int slave, uint32_t index, uint8_t sub, uint8_t value)
 {
   ROS_DEBUG("sdo_bit8: slaveIndex %i, reg 0x%X, subindex %i, value 0x%X", slave, index, sub, value);
-  return ec_SDOwrite(slave, index, sub, FALSE, 1, &value, EC_TIMEOUTRXM);
+  int ReceivedWorkingCounter = ec_SDOwrite(slave, index, sub, FALSE, 1, &value, EC_TIMEOUTRXM);
+  if (ReceivedWorkingCounter == 0)
+  {
+    ROS_FATAL("(sdo_bit8) Error occurred when writing, reg 0x%X, sub-index %i, data: 0x%X to slave %i", index, sub,
+              value, slave);
+  }
+  return ReceivedWorkingCounter;
 }
 
 int sdo_bit16(int slave, uint32_t index, uint8_t sub, uint16_t value)
 {
   ROS_DEBUG("sdo_bit16: slaveIndex %i, reg 0x%X, subindex %i, value 0x%X", slave, index, sub, value);
-  return ec_SDOwrite(slave, index, sub, FALSE, 2, &value, EC_TIMEOUTRXM);
+  int ReceivedWorkingCounter = ec_SDOwrite(slave, index, sub, FALSE, 2, &value, EC_TIMEOUTRXM);
```

Comment:
Working counters are not by definition 0 when nothing has been read/written to the register. The strategy should instead be; read an expected working counter and compare the counter you receive with the one you expect. Only if the counter has not been incremented it should fail. See the following links for more info; https://infosys.beckhoff.com/english.php?content=../content/1033/tc3_io_intro/1257993099.html&id=3196541253205318339
https://github.com/OpenEtherCATsociety/SOEM/blob/master/test/intime/ec_master/ec_master.c

Reply:
I absolutely agree with u to not only check the working on a possible zero value but also thresholds. However my suggestion would be to add these changes in branch:  "fix/PM-196-fix-return-value-initiation-imc-check for testing" . The current check (for 0) is only to assure that an error occurs in writing or reading. If the PDO map is compatible with this version we may add more checks to have a better level of certainty.